numpy/_core/src/umath/stringdtype_ufuncs.cpp

```diff
@@ -0,0 +1,3121 @@
+        &PyArray_StringDType,
+        &PyArray_BoolDType
+    };
+
+    if (init_ufunc(umath, "isnan", bool_output_dtypes,
+                   &string_bool_output_resolve_descriptors,
+                   &string_isnan_strided_loop, 1, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    if (init_ufunc(umath, "isalpha", bool_output_dtypes,
+                   &string_bool_output_resolve_descriptors,
+                   &string_isalpha_strided_loop, 1, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    if (init_ufunc(umath, "isdecimal", bool_output_dtypes,
+                   &string_bool_output_resolve_descriptors,
+                   &string_isdecimal_strided_loop, 1, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    if (init_ufunc(umath, "isdigit", bool_output_dtypes,
+                   &string_bool_output_resolve_descriptors,
+                   &string_isdigit_strided_loop, 1, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    if (init_ufunc(umath, "isnumeric", bool_output_dtypes,
+                   &string_bool_output_resolve_descriptors,
+                   &string_isnumeric_strided_loop, 1, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    if (init_ufunc(umath, "isspace", bool_output_dtypes,
+                   &string_bool_output_resolve_descriptors,
+                   &string_isspace_strided_loop, 1, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    PyArray_DTypeMeta *intp_output_dtypes[] = {
+        &PyArray_StringDType,
+        &PyArray_IntpDType
+    };
+
+    if (init_ufunc(umath, "str_len", intp_output_dtypes,
+                   &string_intp_output_resolve_descriptors,
+                   &string_strlen_strided_loop, 1, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    PyArray_DTypeMeta *binary_dtypes[] = {
+            &PyArray_StringDType,
+            &PyArray_StringDType,
+            &PyArray_StringDType,
+    };
+
+    if (init_ufunc(umath, "maximum", binary_dtypes, binary_resolve_descriptors,
+                   &maximum_strided_loop, 2, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    if (init_ufunc(umath, "minimum", binary_dtypes, binary_resolve_descriptors,
+                   &minimum_strided_loop, 2, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    if (init_ufunc(umath, "add", binary_dtypes, binary_resolve_descriptors,
+                   &add_strided_loop, 2, 1, NPY_NO_CASTING,
+                   (NPY_ARRAYMETHOD_FLAGS) 0) < 0) {
+        return -1;
+    }
+
+    INIT_MULTIPLY(Int8, int8);
```

Comment:
This seems excessive - just let the promotor take care of any types except `Int64` and `UInt64`? Or maybe add `int32`; it's not like those arrays canbe large compared to the string array.

Reply:
Thanks, that refactor totally didn't occur to me.

While implementing this suggestion I also realized that `add` is missing a promoter, so adding with a pystring is broken. Added as a high-prio thing to fix to the issue.